src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SignPostCheck.java

```diff
@@ -195,53 +199,48 @@ public boolean validCheckForObject(final AtlasObject object)
-        int count = 0;
-
         // Go back and collect edges
-        Edge nextEdge = finalEdge;
-        while (count < MAX_EDGE_COUNT_FOR_RAMP)
+        final List<Edge> endEdges = new ArrayList<>();
+        if (maxEdgeCount > 0)
         {
-            count++;
-
-            // Break if edge has no heading
-            final Optional<Heading> initialHeading = nextEdge.asPolyLine().initialHeading();
-            if (!initialHeading.isPresent())
-            {
-                break;
-            }
-
-            // Collect in edges and make sure there is not more than one
-            final Set<Edge> inEdges = nextEdge.inEdges();
-            if (inEdges.isEmpty() || inEdges.size() > 1)
+            // Collect in edges and make sure the list is not empty
+            final Set<Edge> inEdges = finalEdge.inEdges();
+            if (!inEdges.isEmpty())
             {
-                break;
-            }
-
-            // Find the edge that precedes nextEdge
-            final HighwayTag highwayTag = nextEdge.highwayTag();
-            final Edge precedingEdge = inEdges.iterator().next();
-
-            // Ignore if edge has a different classification
-            if (!highwayTag.isIdenticalClassification(precedingEdge.highwayTag()))
-            {
-                break;
-            }
-
-            // Break if given edge has no heading
-            final Optional<Heading> finalHeading = precedingEdge.asPolyLine().finalHeading();
-            if (!finalHeading.isPresent() || initialHeading.get().asPositiveAngle()
-                    .difference(finalHeading.get().asPositiveAngle())
-                    .isGreaterThan(this.rampAngleDifference))
-            {
-                break;
+                // Check each inEdge
+                for (final Edge nextEdge : inEdges)
+                {
+                    // See if it is connected to a source edge
+                    if (this.sourceEdgeFilter.test(nextEdge))
+                    {
+                        endEdges.clear();
+                        return endEdges;
+                    }
+                    // See if it is an arterial
+                    if (!this.rampEdgeFilter.test(nextEdge) && HighwayTag.highwayTag(nextEdge)
+                            .orElse(HighwayTag.NO).isMoreImportantThanOrEqualTo(arterialMinimum))
+                    {
+                        endEdges.clear();
+                        endEdges.add(finalEdge);
+                        return endEdges;
+                    }
+                    // Recurse through inEdges if this edge is not the other side of a bidirectional
+                    // way
+                    else if (nextEdge.highwayTag().isIdenticalClassification(finalEdge.highwayTag())
+                            && finalEdge.getMasterEdgeIdentifier() != nextEdge
+                                    .getMasterEdgeIdentifier())
+                    {
+                        endEdges.addAll(findFirstRampEdge(nextEdge, maxEdgeCount - 1));
```

Comment:
So the idea here is that from the highway edge you walk backwards along all incoming link edges, getting a maximum of maxEdgeCount deep but allowing for multiple branches of that depth?

Reply:
That is correct. This was done to handle branching on ramps and have a hard escape for the recursive loop.